File: web/test/parser.test.ts
import { assert, expect, test } from "vite-plus/test";

import parse from "../src/parse";
import render from "../src/render";
import core from "../src/rulesets/core";
import htmlRenderers from "../src/rulesets/htmlRenderers";

test("basic parse", () => {
	const input = `
# Test ☺️

Here is some text
 *with* bold stuff

* Tight item 1
* Tight item 2

- Loose item 1

- Loose item 2

## Subtest

Here is some more text
`;
	const expected = `
<h1>Test ☺️</h1>
<p>Here is some text
<em>with</em> bold stuff</p>
<ul>
<li>Tight item 1</li>
<li>Tight item 2</li>
</ul>
<ul>
<li>
<p>Loose item 1</p>
</li>
<li>
<p>Loose item 2</p>
</li>
</ul>
<h2>Subtest</h2>
<p>Here is some more text</p>
`.trimStart();

	const doc = parse(input, core);
	const html = render(doc, htmlRenderers);
	expect(html).toBe(expected);

	assert(doc.children);
	expect(doc.children[0].type).toBe("heading");
	expect(doc.children[0].index).toBe(1);
	expect(doc.children[0].length).toBe(9);

	const start = doc.children[0].index;
	const end = start + doc.children[0].length;
	expect(input.substring(start, end)).toBe("# Test ☺️");

	// Let's try it with raw \r
	let input2 = input.replaceAll("\r\n", "\r").replaceAll("\n", "\r");
	const doc2 = parse(input2, core);
	const html2 = render(doc2, htmlRenderers);
	expect(html2.replaceAll("\r\n", "\n").replaceAll("\r", "\n")).toBe(expected);

	assert(doc2.children);
	expect(doc2.children[0].type).toBe("heading");
	expect(doc2.children[0].index).toBe(1);
	expect(doc2.children[0].length).toBe(9);

	const start2 = doc2.children[0].index;
	const end2 = start2 + doc2.children[0].length;
	expect(input2.substring(start2, end2)).toBe("# Test ☺️");
});

File: zig/test/parser_test.zig
const std = @import("std");

const parse = @import("allmark").parse;
const render = @import("allmark").render;
const core = @import("allmark").core;

test "basic parse" {
    const input =
        \\# Test ☺️
        \\
        \\Here is some text
        \\ *with* bold stuff
        \\
        \\* Tight item 1
        \\* Tight item 2
        \\
        \\- Loose item 1
        \\
        \\- Loose item 2
        \\
        \\## Subtest
        \\
        \\Here is some more text
        \\
    ;

    const expected =
        \\<h1>Test ☺️</h1>
        \\<p>Here is some text
        \\<em>with</em> bold stuff</p>
        \\<ul>
        \\<li>Tight item 1</li>
        \\<li>Tight item 2</li>
        \\</ul>
        \\<ul>
        \\<li>
        \\<p>Loose item 1</p>
        \\</li>
        \\<li>
        \\<p>Loose item 2</p>
        \\</li>
        \\</ul>
        \\<h2>Subtest</h2>
        \\<p>Here is some more text</p>
        \\
    ;

    const gpa = std.testing.allocator;
    const rules = try core.init(gpa);
    defer core.deinit(&rules, gpa);

    const doc = try parse.execute(gpa, input, rules);
    defer doc.deinit(gpa);

    const html = try render(gpa, doc, null, false, null);
    defer gpa.free(html);

    try std.testing.expectEqualStrings(expected, html);

    try std.testing.expect(doc.children != null);
    try std.testing.expectEqualStrings("heading", doc.children.?[0].type);
    try std.testing.expectEqual(@as(usize, 0), doc.children.?[0].index);
    try std.testing.expectEqual(@as(usize, 13), doc.children.?[0].length);

    const start = doc.children.?[0].index;
    const end = start + doc.children.?[0].length;
    try std.testing.expectEqualStrings("# Test ☺️", input[start..end]);

    const input2 = try std.mem.replaceOwned(u8, gpa, input, "\r\n", "\r");
    defer gpa.free(input2);
    const input2b = try std.mem.replaceOwned(u8, gpa, input2, "\n", "\r");
    defer gpa.free(input2b);

    const doc2 = try parse.execute(gpa, input2b, rules);
    defer doc2.deinit(gpa);

    const html2 = try render(gpa, doc2, null, false, null);
    defer gpa.free(html2);

    const normalized_html2a = try std.mem.replaceOwned(u8, gpa, html2, "\r\n", "\n");
    defer gpa.free(normalized_html2a);
    const normalized_html2b = try std.mem.replaceOwned(u8, gpa, normalized_html2a, "\r", "\n");
    defer gpa.free(normalized_html2b);

    try std.testing.expectEqualStrings(expected, normalized_html2b);

    try std.testing.expect(doc2.children != null);
    try std.testing.expectEqualStrings("heading", doc2.children.?[0].type);
    try std.testing.expectEqual(@as(usize, 0), doc2.children.?[0].index);
    try std.testing.expectEqual(@as(usize, 13), doc2.children.?[0].length);

    const start2 = doc2.children.?[0].index;
    const end2 = start2 + doc2.children.?[0].length;
    try std.testing.expectEqualStrings("# Test ☺️", input2b[start2..end2]);
}
